dot/core/digest.go

```diff
@@ -0,0 +1,304 @@
+	auths   []*types.GrandpaAuthorityData
+	atBlock *big.Int
+}
+
+type pause struct {
+	atBlock *big.Int
+}
+
+type resume struct {
+	atBlock *big.Int
+}
+
+func newDigestHandler(blockState BlockState, babe BlockProducer, grandpa FinalityGadget) *digestHandler {
+	return &digestHandler{
+		blockState: blockState,
+		grandpa:    grandpa,
+		babe:       babe,
+		stopped:    true,
+	}
+}
+
+func (h *digestHandler) start() {
+	go h.handleGrandpaChanges()
+	h.stopped = false
+}
+
+func (h *digestHandler) stop() {
+	h.stopped = true
+}
+
+func (h *digestHandler) handleGrandpaChanges() {
+	for {
+		if h.stopped {
+			return
+		}
+
+		// TODO: update block state to register new channels for added blocks
+		curr, err := h.blockState.BestBlockHeader()
+		if err != nil {
+			continue
+		}
+
+		pause := h.grandpaPause
+		if pause != nil && curr.Number.Cmp(pause.atBlock) == 0 {
+			// save authority data for Resume
+			h.grandpaAuths = h.grandpa.Authorities()
+			h.grandpa.UpdateAuthorities([]*types.GrandpaAuthorityData{})
+			h.grandpaPause = nil
+			continue
+		}
+
+		resume := h.grandpaResume
+		if resume != nil && curr.Number.Cmp(resume.atBlock) == 0 {
+			h.grandpa.UpdateAuthorities(h.grandpaAuths)
+			h.grandpaResume = nil
+			continue
+		}
+
+		fc := h.grandpaForcedChange
+		if fc != nil && curr.Number.Cmp(fc.atBlock) == 0 {
+			h.grandpa.UpdateAuthorities(fc.auths)
+			h.grandpaForcedChange = nil
+			continue
+		}
+
+		sc := h.grandpaScheduledChange
+		if sc != nil && curr.Number.Cmp(sc.atBlock) == 0 {
+			h.grandpa.UpdateAuthorities(sc.auths)
+			h.grandpaScheduledChange = nil
+		}
+
+		time.Sleep(time.Millisecond * 100)
+	}
+}
+
+// handleConsensusDigest is the function used by the syncer to handler a consensus digest
+func (h *digestHandler) handleConsensusDigest(d *types.ConsensusDigest) error {
+	t := d.DataType()
+
+	switch t {
+	case types.ScheduledChangeType:
+		return h.handleScheduledChange(d)
+	case types.ForcedChangeType:
+		return h.handleForcedChange(d)
+	case types.OnDisabledType:
+		return h.handleOnDisabled(d)
+	case types.PauseType:
+		return h.handlePause(d)
+	case types.ResumeType:
+		return h.handleResume(d)
+	default:
+		return errors.New("invalid consensus digest data")
+	}
+}
+
+func (h *digestHandler) handleScheduledChange(d *types.ConsensusDigest) error {
+	curr, err := h.blockState.BestBlockHeader()
+	if err != nil {
+		return err
+	}
+
+	if d.ConsensusEngineID == types.BabeEngineID {
+		// TODO
+	} else {
+		if h.grandpaScheduledChange != nil {
+			return errors.New("already have scheduled change scheduled")
+		}
+
+		sc := &types.GrandpaScheduledChange{}
+		dec, err := scale.Decode(d.Data[1:], sc)
+		if err != nil {
+			return err
+		}
+		sc = dec.(*types.GrandpaScheduledChange)
+
+		c, err := newGrandpaChange(sc.Auths, sc.Delay, curr.Number)
+		if err != nil {
+			return err
+		}
+
+		h.grandpaScheduledChange = c
+	}
+
+	return nil
+}
+
+func (h *digestHandler) handleForcedChange(d *types.ConsensusDigest) error {
```

Comment:
I don't understand the difference between this and handleScheduledChange?  Is it that one deals with finalized blocks, while the other deals with imported blocks?

Reply:
yeah, I believe that is correct, `ForcedChange` is for imported and `ScheduleChange` is for finalized blocks. it also says a `ForcedChange` should be disregarded if blocks have been finalized before it executes, but there isn't functionality for receiving a finalized block directly from the state service yet, I'll open an issue for that and to better define these 2 types